scripts/ci/ws12_workflow_contracts.py

```
@@ -1741,6 +1743,173 @@ def validate_crate_name_residue(
+    stays scoped to the one command step; matrix expressions used for display
+    names and artifact paths remain valid workflow metadata.
+    """
+
+    text = workflows.get(ROOT_TEST_WORKFLOW)
+    if text is None:
+        return []
+    job = job_body(text, ROOT_TEST_JOB)
+    if job is None:
+        return [
+            f"{ROOT_TEST_WORKFLOW}: missing {ROOT_TEST_JOB!r} job, so the root "
+            "test command cannot be checked for direct matrix interpolation"
```

Ruff (0.16.2)

[warning] Unparenthesized implicit string concatenation in collection

Did you forget a comma?

(ISC004)